#2794 - See GenMapper On Every Group Record

kodinkat opened the pull request:
- fixes: #2794 
---

<img width="757" height="1069" alt="Screenshot 2025-11-25 at 17 16 40" src="https://github.com/user-attachments/assets/3eafc945-9275-4b38-b9ec-2b6cb4a31951" />

kodinkat committed:
Add generational map tile functionality

                - Introduced a new tile for displaying a generational map in the groups section.
                - Added JavaScript for fetching and rendering the generational map.
                - Updated PHP files to include hooks and display logic for the new tile.
                - Enhanced the existing tile display logic to accommodate the new 'genmap' tile.
                - Enqueued necessary scripts and styles for the generational map functionality.

kodinkat committed:
Add group generational map details and selection functionality

                - Implemented a new footer action for group records to display metrics modals.
                - Enhanced JavaScript to manage node selection and display detailed information for selected nodes in the generational map.
                - Added CSS for visual indication of selected nodes.
                - Updated PHP to ensure correct naming for groups in the generated map.
                - Introduced modal functionality for adding child groups with appropriate UI elements and event handling.

kodinkat committed:
Enhance generational map functionality and layout options

                - Added a new action to render the generational map after tiles in the single template.
                - Updated the JavaScript to support layout toggling between vertical and horizontal views, saving user preferences in local storage.
                - Improved the rendering logic to prevent double rendering of the generational map.
                - Enhanced CSS for better responsiveness and visual clarity in both layouts.
                - Introduced a modal for displaying detailed information about selected nodes in the generational map.

kodinkat committed:
Merge branch 'develop' of https://github.com/DiscipleTools/disciple-tools-theme into 2794-see-genmapper-on-every-group

kodinkat committed:
Enhance D3.js Genmap Visualization

                - Introduced D3.js for the new generational map visualization, replacing the legacy jQuery OrgChart implementation.
                - Added support for group type icons and status colors in the D3 visualization.
                - Implemented layout toggle functionality for responsive design.
                - Updated SQL queries to include group_type field for groups post type.
                - Added CSS styles for D3.js visualization and popover interactions.

                This commit marks a significant step towards modernizing the visualization component and improving user experience.

kodinkat committed:
Update SQL query in Genmap class to include safe handling for group_type field

                - Added a comment to ignore PHPCS warning for prepared SQL fragments related to group_type.
                - Ensured that the SQL query remains secure while incorporating the necessary group_type logic.

kodinkat committed:
Refactor SQL query preparation in Genmap class

                - Removed redundant PHPCS disable comment and ensured consistent code style.
                - Maintained secure handling of SQL queries while preparing for group type logic.

corsacca commented:
Nice work @kodinkat!

Here are some Notes/questions. These aren't decisions yet, just things we need to connect on as a group.
- in the 2 column mode, the rest of the tiles don't distribute. they are all in the same column.
- We might consider disabling scrolling, It can make the page navigation harder. 
- All groups are created as a pregroup. We might want more fields when adding a child group.
- We always display the root group as gen zero, This might be confusing, As we might be dealing with a second generation group.
 - Do we want to display any health icons on the node? 
 
 @cairocoder01 your thoughts or more questions welcome

corsacca commented:
Thanks @cairocoder01 
Consensus todos:
- default single column
- upgrade icon
- explore zoom controls
- remove gen numbers

Zoom. I've seen this on another project: scroll could be disabled until the user clicks into the canvas. This reduces a bit of the frustration when trying to scroll past the GenMapper, and still keeps the scroll functionality to go in and out. 

Connecting to an existing group. Would this be from clicking on a record with a separate button? or would the add model have the option to connect to an existing group? 

More discussion needed on group health icons and displaying the group type. What happens with custom group health option and custom group types?

kodinkat committed:
Merge branch 'develop' of https://github.com/DiscipleTools/disciple-tools-theme into 2794-see-genmapper-on-every-group

kodinkat committed:
Enhance generational map functionality with layout adjustments and modal support

                - Updated the genmap section to support responsive layouts with new CSS classes for better display on different screen sizes.
                - Added a maximize button to open the generational map in a fullscreen modal, improving user experience.
                - Implemented logic to determine graph orientation based on section classes, allowing for dynamic rendering of the graph layout.
                - Enhanced D3 rendering functions to accommodate both vertical and horizontal layouts, ensuring proper visualization based on the current context.
                - Introduced collapse indicators for nodes with children, improving interactivity within the graph.
                - Refactored existing JavaScript functions to streamline the rendering process and improve performance.

kodinkat committed:
Refactor graph orientation logic and enhance rendering for generational map

                - Updated default graph orientation to vertical for better initial display.
                - Adjusted comments and logic to clarify the relationship between section classes and graph orientation.
                - Enhanced SVG rendering to support dynamic rotation based on graph orientation.
                - Improved link generation for both horizontal and vertical layouts, ensuring cleaner connections between nodes.
                - Streamlined the process of determining graph orientation in modal contexts, enhancing overall user experience.

kodinkat committed:
Refactor whitespace and formatting in genmap-tile.js for improved readability

                - Cleaned up unnecessary whitespace and adjusted formatting for consistency throughout the file.
                - Enhanced clarity in the graph orientation logic and link path calculations.
                - Ensured better alignment of comments with the corresponding code sections.

kodinkat committed:
Refactor link connection logic in genmap-tile.js for consistent node alignment

                                - Simplified the logic for connecting nodes in both horizontal and vertical layouts to ensure consistent alignment with tree hierarchy.
                                - Removed unnecessary calculations for node alignment, enhancing code clarity and maintainability.
                                - Updated comments to reflect the new approach for connecting parent and child nodes.

corsacca commented on single-template.php:
```diff
@@ -303,9 +303,13 @@ function dt_display_tile( $tile, $post ): bool {
                     <div class="cell small-12">
                         <div class="grid-x grid-margin-x grid-margin-y grid">
                             <?php
+                            // Render genmap first if it exists
```
@kodinkat this hook will now be called twice. Any plugin that uses this hook will have content rendered twice.
What is the catch with registering the tile the normal way and hooking into dt_details_additional_section for displaying the content?

Registering the tile the normal way will also let admin hide the tile if they want. And we can provide a description shown when the user click the help icon.

kodinkat committed:
Refactor generational map rendering and tile display logic

                - Removed legacy genmap rendering logic from single-template.php and base-setup.php to streamline the display process.
                - Updated the genmap tile JavaScript to enhance D3.js visualization, ensuring a more responsive layout and improved user interaction.
                - Simplified the conditions for rendering tiles, focusing on essential elements for better performance and maintainability.

corsacca commented on single-template.php:
```diff
@@ -377,8 +373,7 @@ function dt_display_tile( $tile, $post ): bool {
                                 </section>
-                            <?php }
-                            do_action( 'dt_record_bottom_after_tiles', $post_type, $dt_post ); ?>
+                            <?php } ?>
```
@kodinkat don't deleted the original dt_record_bottom_after_tiles

kodinkat committed:
Enhance genmap initialization and rendering logic

                                - Added logic to initialize the genmap only for the 'groups' post type, ensuring it loads correctly when the DOM is ready.
                                - Updated the genmap tile JavaScript to include an idempotent initialization function, preventing multiple initializations and improving performance.
                                - Simplified the conditions for rendering the genmap tile, ensuring it is always rendered while controlling visibility through CSS.
                                - Improved comments for clarity and maintainability of the code.

kodinkat committed:
Update single-template.php to include new action hooks for tile display

                - Reverted back to action hook 'dt_record_bottom_after_tiles'.

corsacca commented:
Thanks @kodinkat! And thanks for the reviews @cairocoder01!
